packages/core/src/sync/index.ts

```diff
@@ -1,318 +1,592 @@
+          from,
+          to: to < _estimate ? to : _estimate,
+          limit,
         });
-      return _block as SyncBlock;
-    });
-
-/**
- * Helper function for "eth_getLogs" rpc request.
- * Handles different error types and retries the request if applicable.
- */
-export const _eth_getLogs = async (
-  { requestQueue }: Pick<BaseSyncService, "requestQueue">,
-  params: {
-    address?: Address | Address[];
-    topics?: LogTopic[];
-  } & (
-    | { fromBlock: Hex | number; toBlock: Hex | number }
-    | { blockHash: Hash }
-  ),
-): Promise<SyncLog[]> => {
-  if ("blockHash" in params) {
-    return requestQueue
-      .request({
-        method: "eth_getLogs",
-        params: [
+
+        updateStatus(events, cursor, false);
+
+        // Use range and number of events returned to update estimate
+        // 10 <= estimate(new) <= estimate(prev) * 2
+        estimateSeconds = Math.min(
+          Math.max(
+            10,
+            Math.round(
+              (limit * decodeCheckpoint(cursor).blockTimestamp -
+                decodeCheckpoint(from).blockTimestamp) /
+                (events.length || 1),
+            ),
+          ),
+          estimateSeconds * 2,
+        );
+
+        yield { events, checkpoint: to };
+        from = cursor;
+      }
+
+      // Exit condition: All network have completed historical sync.
+      if (
+        _localSyncs.every(
+          (localSync) =>
+            localSync.isComplete() ||
+            localSync.finalizedBlock === localSync.latestBlock,
+        )
+      ) {
+        break;
+      }
+    }
+  }
+
+  /**
+   * Omnichain `onRealtimeSyncEvent`
+   *
+   * Handle callback events across all `args.networks`, and raising these
+   * events to `args.onRealtimeEvent` while maintaining checkpoint ordering.
+   *
+   * Note: "block" events are still being handled by writing and reading from
+   * the sync-store. This approach is not future proof and inefficient.
+   */
+  const eventQueue = createQueue({
+    browser: false,
+    concurrency: 1,
+    initialStart: true,
+    worker: async ({
+      network,
+      event,
+    }: { network: Network; event: RealtimeSyncEvent }) => {
+      const localSync = localSyncs.get(network)!;
+      const realtimeSync = realtimeSyncs.get(network)!;
+      switch (event.type) {
+        /**
+         * Handle a new block being ingested.
+         */
+        case "block":
+          {
+            const filters = args.sources
+              .filter(({ filter }) => filter.chainId === network.chainId)
+              .map(({ filter }) => filter);
+
+            // Update local sync, record checkpoint before and after
+            let from = getChainsCheckpoint("latest")!;
+            localSync.latestBlock = event.block;
+            const to = getChainsCheckpoint("latest")!;
+
+            // Add block, logs, transactions, receipts, and traces to the sync-store.
+
+            const chainId = network.chainId;
+
+            await Promise.all([
```

Comment:
Awesome that this is all in one place now, clear path to better batching, in-memory tricks, etc